src/v/cloud_storage/inventory/report_parser.cc

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "cloud_storage/inventory/report_parser.h"
+
+#include "base/vassert.h"
+
+#include <seastar/util/short_streams.hh>
+
+#include <zlib.h>
+
+namespace {
+constexpr auto newline{'\n'};
+} // namespace
+
+namespace cloud_storage::inventory {
+
+report_parser::report_parser(
+  ss::input_stream<char> stream,
+  gzip_compressed_t is_compressed,
+  size_t chunk_size)
+  : _stream{std::move(stream)}
+  , _is_compressed{is_compressed}
+  , _chunk_size{chunk_size} {}
+
+ss::future<std::optional<ss::sstring>> report_parser::next_line() {
+    while (!_buffer.contains(newline)) {
```

Comment:
that could be zero copy with iobuf

Reply:
I switched to use iobuf in the stream decompression code (`gzip_stream_decompressor`). We now also return at most `chunk_size` (passed as ctor arg to decompression obj) per decompression call and keep state around between calls to `gzip_stream_decompressor::next` to keep the size under limit. The chunk size by default is set to 128KiB.

However I did not change the type in the report parser. It still needs to work with strings to detect newline, and it has to buffer data until it finds a newline in the csv file.

It now calls into the `gzip_stream_decompressor::next` and receives 128KiB of decompressed data at most in one call, and then converts that iobuf to string and extracts rows out of it.

Since the max key name in AWS is 1024 bytes and bucket name is max 63 bytes, one call should get us several rows at once (1024+63+few bytes for quotes and newlines). 

One problem that I see now we might run into is the string encoding. For the paths we always use ascii character set, but there is no guarantee that we won't have any utf-8 encoded paths in the bucket, since AWS allows any valid UTF-8 as keys. If such an object is placed in the bucket, it will be part of the inventory report, and while parsing we might try to convert it to `sstring` and get some unexpected results.

Since filtering rows for NTP should remove such rows from further processing, we should be safe, but it might be something to keep in mind.

Reply:
>However I did not change the type in the report parser. It still needs to work with strings to detect newline, and it has to buffer data until it finds a newline in the csv file.

can't we detect newline in an iobuf?

Reply:
I've made changes to use iobuf internally within the report parser, it outputs a fragmented vector of strings as the individual rows are expected to be small. 

I'm not sure if splitting the iobuf on newline was written in the best way possible here though. Perhaps there is a better approach compared to what I used (the bytes iterator is used on an iobuf to extract rows out of it).